prog/vnet/aws/vpc_nexus.rb

```diff
@@ -0,0 +1,193 @@
+      private_subnet.private_subnet_aws_resource.update(internet_gateway_id:)
+      if internet_gateway.attachments.empty?
+        client.attach_internet_gateway({internet_gateway_id:, vpc_id: private_subnet.private_subnet_aws_resource.vpc_id})
+      end
+    end
+
+    begin
+      client.create_route({
+        route_table_id: route_table_id,
+        destination_ipv_6_cidr_block: "::/0",
+        gateway_id: internet_gateway_id
+      })
+
+      client.create_route({
+        route_table_id: route_table_id,
+        destination_cidr_block: "0.0.0.0/0",
+        gateway_id: internet_gateway_id
+      })
+    rescue Aws::EC2::Errors::RouteAlreadyExists
+    end
+
+    hop_wait
+  end
+
+  label def wait
+    when_update_firewall_rules_set? do
+      private_subnet.vms.each(&:incr_update_firewall_rules)
+      decr_update_firewall_rules
+    end
+
+    private_subnet.semaphores.each(&:destroy)
+
+    nap 60 * 60 * 24 * 365
+  end
+
+  label def destroy
+    if private_subnet.nics.any? { |n| !n.vm_id.nil? }
+      register_deadline(nil, 10 * 60, allow_extension: true) if private_subnet.nics.any? { |n| n.vm&.prevent_destroy_set? }
+
+      Clog.emit("Cannot destroy subnet with active nics, first clean up the attached resources") { private_subnet }
+
+      nap 5
+    end
+
+    decr_destroy
+    private_subnet.nics.each(&:incr_destroy)
+    private_subnet.firewalls.each(&:destroy)
+    Semaphore.incr(strand.children_dataset.where(prog: "Aws::Vpc").select(:id), "destroy")
+
+    ignore_invalid_id do
+      client.delete_security_group({group_id: private_subnet.private_subnet_aws_resource.security_group_id})
+    end
+    hop_delete_internet_gateway
+  end
+
+  label def delete_internet_gateway
+    ignore_invalid_id do
+      client.detach_internet_gateway({internet_gateway_id: private_subnet.private_subnet_aws_resource.internet_gateway_id, vpc_id: private_subnet.private_subnet_aws_resource.vpc_id})
+      client.delete_internet_gateway({internet_gateway_id: private_subnet.private_subnet_aws_resource.internet_gateway_id})
+    end
+    hop_delete_vpc
+  end
+
+  label def delete_vpc
+    ignore_invalid_id do
+      client.delete_vpc({vpc_id: private_subnet.private_subnet_aws_resource.vpc_id})
+    end
+
+    nap 5 unless private_subnet.nics.empty?
+    private_subnet.private_subnet_aws_resource.destroy
+    private_subnet.destroy
+    pop "vpc destroyed"
+  end
+
+  def ignore_invalid_id
+    yield
+  rescue ArgumentError,
+    Aws::EC2::Errors::GatewayNotAttached,
+    Aws::EC2::Errors::InvalidSubnetIDNotFound,
+    Aws::EC2::Errors::InvalidGroupNotFound,
+    Aws::EC2::Errors::InvalidNetworkInterfaceIDNotFound,
+    Aws::EC2::Errors::InvalidInternetGatewayIDNotFound,
+    Aws::EC2::Errors::InvalidVpcIDNotFound => e
+    Clog.emit("ID not found for aws vpc") { {ignored_aws_vpc_failure: {exception: Util.exception_to_hash(e, backtrace: nil)}} }
+  end
+
+  def location
+    private_subnet.location
```

Comment:
Up to you, but you may want to cache this, since the location shouldn't change:

```suggestion
    @location ||= private_subnet.location
```